libkbfs/block_retrieval_worker.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"io"
+	"reflect"
+
+	"golang.org/x/net/context"
+)
+
+// blockRetrievalWorker processes blockRetrievalQueue requests
+type blockRetrievalWorker struct {
+	blockGetter
+	stopCh chan struct{}
+	queue  *blockRetrievalQueue
+}
+
+// run runs the worker loop until Shutdown is called
+func (brw *blockRetrievalWorker) run() {
+	for {
+		select {
+		case <-brw.stopCh:
+			return
+		default:
+			brw.HandleRequest()
+		}
+	}
+}
+
+// newBlockRetrievalWorker returns a blockRetrievalWorker for a given
+// blockRetrievalQueue, using the passed in blockGetter to obtain blocks for
+// requests.
+func newBlockRetrievalWorker(bg blockGetter, q *blockRetrievalQueue) *blockRetrievalWorker {
+	brw := &blockRetrievalWorker{
+		blockGetter: bg,
+		stopCh:      make(chan struct{}),
+		queue:       q,
+	}
+	go brw.run()
+	return brw
+}
+
+// notifyBlockRequestor copies the source block into the request's block
+// pointer, and notifies the channel that the request is waiting on. Should be
+// called in a goroutine.
+func notifyBlockRequestor(req *blockRetrievalRequest, source reflect.Value, err error) {
+	// Copy the decrypted block to the caller
+	dest := reflect.ValueOf(req.block).Elem()
+	dest.Set(source)
+	req.doneCh <- err
+}
+
+// HandleRequest is the main work method for the worker. It obtains a
+// blockRetrieval from the queue, retrieves the block using
+// blockGetter.getBlock, and responds to the subscribed requestors with the
+// results.
+func (brw *blockRetrievalWorker) HandleRequest() (err error) {
+	retrievalCh := brw.queue.WorkOnRequest()
+	var retrieval *blockRetrieval
+	select {
+	case retrieval = <-retrievalCh:
+	case <-brw.stopCh:
+		return io.EOF
+	}
+
+	// Create a new block of the same type as the first request
+	typ := reflect.TypeOf(retrieval.requests[0].block).Elem()
+	block := reflect.New(typ).Interface().(Block)
+	defer func() {
+		brw.queue.FinalizeRequest(retrieval, block, err)
+	}()
+
+	// Pick one of the still-active contexts to use
```

Comment:
Hrm, I didn't think about this.  It's somewhat annoying that the picked context could cancel, and other waiters would get a canceled error even when their context is still good.  Probably not worth fixing, but you can imagine building some megacontext that only gets canceled when all the request contexts are canceled, and combines the debug tags from all requests as well.  Maybe a TODO?